types/wlr_seat.c

```diff
@@ -819,3 +889,259 @@ void wlr_seat_keyboard_notify_key(struct wlr_seat *seat, uint32_t time,
+
+	return point;
+}
+
+struct wlr_touch_point *wlr_seat_touch_get_point(
+		struct wlr_seat *seat, int32_t touch_id) {
+	struct wlr_touch_point *point = NULL;
+	wl_list_for_each(point, &seat->touch_state.touch_points, link) {
+		if (point->touch_id == touch_id) {
+			return point;
+		}
+	}
+
+	return NULL;
+}
+
+void wlr_seat_touch_notify_down(struct wlr_seat *seat,
+		struct wlr_surface *surface, uint32_t time, int32_t touch_id, double sx,
+		double sy) {
+	clock_gettime(CLOCK_MONOTONIC, &seat->last_event);
+	struct wlr_seat_touch_grab *grab = seat->touch_state.grab;
+	struct wlr_touch_point *point =
+		touch_point_create(seat, touch_id, surface, sx, sy);
+	if (!point) {
+		wlr_log(L_ERROR, "could not create touch point");
+		return;
+	}
+
+	grab->interface->down(grab, time, point);
+
+	if (wl_list_length(&seat->touch_state.touch_points) == 1) {
+		seat->touch_state.grab_serial = wl_display_get_serial(seat->display);
+		seat->touch_state.grab_id = touch_id;
+	}
+}
+
+void wlr_seat_touch_notify_up(struct wlr_seat *seat, uint32_t time,
+		int32_t touch_id) {
+	clock_gettime(CLOCK_MONOTONIC, &seat->last_event);
+	struct wlr_seat_touch_grab *grab = seat->touch_state.grab;
+	struct wlr_touch_point *point = wlr_seat_touch_get_point(seat, touch_id);
+	if (!point) {
+		wlr_log(L_ERROR, "got touch up for unknown touch point");
+		return;
+	}
+
+	grab->interface->up(grab, time, point);
+
+	touch_point_destroy(point);
+}
+
+void wlr_seat_touch_notify_motion(struct wlr_seat *seat, uint32_t time,
+		int32_t touch_id, double sx, double sy) {
+	clock_gettime(CLOCK_MONOTONIC, &seat->last_event);
+	struct wlr_seat_touch_grab *grab = seat->touch_state.grab;
+	struct wlr_touch_point *point = wlr_seat_touch_get_point(seat, touch_id);
+	if (!point) {
+		wlr_log(L_ERROR, "got touch motion for unknown touch point");
+		return;
+	}
+
+	point->sx = sx;
+	point->sy = sy;
+
+	grab->interface->motion(grab, time, point);
+}
+
+static void touch_point_clear_focus(struct wlr_touch_point *point) {
+	if (point->focus_surface) {
+		wl_list_remove(&point->focus_surface_destroy.link);
+		point->focus_client = NULL;
+		point->focus_surface = NULL;
+	}
+}
+
+static void handle_point_focus_destroy(struct wl_listener *listener,
+		void *data) {
+	struct wlr_touch_point *point =
+		wl_container_of(listener, point, focus_surface_destroy);
+	touch_point_clear_focus(point);
+}
+
+static void touch_point_set_focus(struct wlr_touch_point *point,
+		struct wlr_surface *surface, double sx, double sy) {
+	if (point->focus_surface == surface) {
+		return;
+	}
+
+	touch_point_clear_focus(point);
+
+	if (surface && surface->resource) {
+		struct wlr_seat_client *client =
+			wlr_seat_client_for_wl_client(point->client->seat,
+				wl_resource_get_client(surface->resource));
+
+		if (client && client->touch) {
+			wl_signal_add(&surface->events.destroy, &point->focus_surface_destroy);
+			point->focus_surface_destroy.notify = handle_point_focus_destroy;
```

Comment:
This could be set in `create`, once.

Reply:
I've adopted this style from Weston to always set the notify function within one line of the `wl_signal_add()` call and it's used all over the project.